app/services/katello/applicable_host_queue.rb

```diff
@@ -0,0 +1,21 @@
+module Katello
+  class ApplicableHostQueue
+    def self.batch_size
+      ::Setting::Content.find_by(name: "applicability_batch_size").value
+    end
+
+    def self.queue_depth
+      ::Katello::HostQueueElement.all.size
+    end
+
+    def self.push_host(host_id)
+      HostQueueElement.create!({ host_id: host_id })
+    end
+
+    def self.pop_hosts(amount = self.batch_size)
+      queue = HostQueueElement.order(:created_at => 'asc')[0...amount]
+      HostQueueElement.delete(queue)
+      queue
```

Comment:
i think this won't de-duplicate host queue element items, if i'm reading it right.  

Also, in AR .limit(amount) would be a lot faster than using array methods [0..amount] (i think)

Reply:
Do you think a good strategy for removing the duplicates would be to retrieve the popped hosts, delete them from the DB, and then return the batched host list with a `.uniq`?

Reply:
I think so, i'd probably do something like:
HostQueueElement.group(:host_id).select("host_id, MIN(created_at) as created_at").limit(amount)
and then you can just delete_all by host_id and pass the results out